docs/proposals/2779-trainjob-progress/README.md

```diff
@@ -0,0 +1,656 @@
+
+I can use standard tools, like the Kubeflow Trainer Python SDK or `kubectl get trainjob` to access progress and performance metrics about my train jobs.
+
+### Story 3: Data Scientist / ML Engineer opting out of monitoring training jobs
+
+As a data scientist or ML Engineer, I want to create a TrainJob but I do not want to have to work out how to integrate training monitoring.
+
+## Proposal option 1: Push-based
+
+As a first option, we propose a **push-based** approach with the following high-level design:
+
+1. The TrainJob custom resource exposes the current training progress and metrics via a new optional field `status.trainerStatus`.
+2. The trainer control plane exposes a new http service which can receive the trainer status from the trainer runtime pods.
+3. The user instruments their trainer runtime pod(s) to periodically send the current trainer status to this endpoint which updates the status of the TrainJob.
+
+The feature is optional but available for all TrainJobs. Users opt in to the functionality by instrumenting their runtime to send requests to the new endpoint.
+
+Support for any ML training framework is enabled by instrumentation code tailored to each framework. Users can provide their own instrumentation code, though to make it easier for users to adopt the feature, we propose:
+
+4. Adding new built-in trainers to the kubeflow-sdk that automatically inject the instrumentation code for common ML frameworks. Initially we propose only adding a builtin trainer for the [*Transformers* *Trainer* API](https://huggingface.co/docs/transformers/en/main_classes/trainer).
+
+### Design Details
+
+#### CRD changes
+
+The TrainJob API would be updated to include a new optional `status.trainerStatus` field with this schema:
+
+```go
+type TrainJobStatus struct {
+    // ... existing fields
+
+    // trainerStatus provides a summary of the status of the training 
+	// part of the TrainJob.
+    // Empty if the status is unknown, e.g. the job has just started
+    // or the job is not instrumented to report its status.
+    // +optional
+    TrainerStatus *TrainJobTrainerStatus `json:"trainerStatus,omitempty"`
+}
+
+
+type TrainJobTrainerStatus struct {
+
+    // progressPercentage gives an estimate of how complete the TrainJob is as a percentage.
+    // The value will be between 0 and 100, or empty if unknown.
+    //
+    // +kubebuilder:validation:Minimum=0
+    // +kubebuilder:validation:Maximum=100
+    // +optional
+    ProgressPercentage *int32 `json:"progressPercentage,omitempty"`
+
+    // estimatedRemainingSeconds gives the estimated remaining training time in seconds
+    // before the train job is completed.
+    // The value will be empty if it is unknown.
+    //
+    // +kubebuilder:validation:Minimum=0
+    // +optional
+    EstimatedRemainingSeconds *int64 `json:"estimatedRemainingSeconds,omitempty"`
```

Comment:
How would we calculate this? Will we use approach that checks the time of the first epoch, and get the total time as: first_epoch_time * num_epochs.

I found out that the first epoch might include data decompression, dataset shuffling/caching, or copying to GPU caches. Which might be slower than the rest of the training process.

Another idea could be to take medium for per batch time.

We can explore some papers in this area, like:
- https://www.microsoft.com/en-us/research/wp-content/uploads/2021/02/dnnperf.pdf
- https://www.research.unipd.it/retrieve/e14fb26f-831e-3de1-e053-1705fe0ac030/NeurIPS-2020-predicting-training-time-without-training-Paper.pdf